src/decoding.rs

```diff
@@ -259,14 +219,48 @@ pub fn decode<T: DeserializeOwned>(
     key: &DecodingKey,
     validation: &Validation,
 ) -> Result<TokenData<T>> {
-    match verify_signature(token, key, validation) {
-        Err(e) => Err(e),
-        Ok((header, claims)) => {
-            let decoded_claims = DecodedJwtPartClaims::from_jwt_part_claims(claims)?;
-            let claims = decoded_claims.deserialize()?;
-            validate(decoded_claims.deserialize()?, validation)?;
-
-            Ok(TokenData { header, claims })
+    // The decoding step is unfortunately a little clunky but that seems to be how I need to do it to fit it into the `decode` function
```

Comment:
how would it look like without that constraint?

Reply:
Hmm, I think that comment doesn't really apply anymore (I should have removed it). There was a stage that I was less sure what was going on so I had a bit of a mess in there :facepalm: